pennylane/devices/qubit/adjoint_jacobian.py

```diff
@@ -108,3 +116,138 @@ def adjoint_jacobian(tape: QuantumTape):  # pylint: disable=too-many-statements
+    if set(tape.wires) != set(range(tape.num_wires)):
+        wire_map = {w: i for i, w in enumerate(tape.wires)}
+        tape = qml.map_wires(tape, wire_map)
+
+    ket = _get_output_ket(tape)
+
+    n_obs = len(tape.observables)
+    bras = np.empty([n_obs] + [2] * len(tape.wires), dtype=np.complex128)
+    for i, obs in enumerate(tape.observables):
+        bras[i] = apply_operation(obs, ket)
+
+    param_number = len(tape.get_parameters(trainable_only=False, operations_only=True)) - 1
+    trainable_param_number = len(tape.trainable_params) - 1
+
+    tangents_out = np.zeros(n_obs)
+
+    for op in reversed(tape._ops):
+        adj_op = qml.adjoint(op)
+        ket = apply_operation(adj_op, ket)
+
+        if op.grad_method is not None:
+            if param_number in tape.trainable_params:
+                # don't do anything if the tangent is 0
+                if not qml.math.allclose(tangents[trainable_param_number], 0):
+                    d_op_matrix = operation_derivative(op)
+                    ket_temp = apply_operation(qml.QubitUnitary(d_op_matrix, wires=op.wires), ket)
+
+                    tangents_out += (
+                        2
+                        * _dot_product_real(bras, ket_temp, len(tape.wires))
+                        * tangents[trainable_param_number]
+                    )
+
+                trainable_param_number -= 1
+            param_number -= 1
+
+        for i in range(n_obs):
+            bras[i] = apply_operation(adj_op, bras[i])
+
+    if n_obs == 1:
+        return np.array(tangents_out[0])
+
+    return tuple(np.array(t) for t in tangents_out)
+
+
+def adjoint_vjp(tape: QuantumTape, cotangents: Tuple[Number]):
+    """Implements the adjoint method outlined in
+    `Jones and Gacon <https://arxiv.org/abs/2009.02823>`__ to differentiate an input tape.
+
+    After a forward pass, the circuit is reversed by iteratively applying adjoint
+    gates to scan backwards through the circuit.
+
+    .. note::
+        The adjoint differentiation method has the following restrictions:
+
+        * Only expectation values are supported as measurements.
+
+        * Cannot differentiate with respect to observables.
+
+        * Observable being measured must have a matrix.
+
+    Args:
+        tape (.QuantumTape): circuit that the function takes the gradient of
+
+    Returns:
+        array or tuple[array]: the derivative of the tape with respect to trainable parameters.
+        Dimensions are ``(len(observables), len(trainable_params))``.
+    """
+    # Map wires if custom wire labels used
+    if set(tape.wires) != set(range(tape.num_wires)):
+        wire_map = {w: i for i, w in enumerate(tape.wires)}
+        tape = qml.map_wires(tape, wire_map)
+
+    ket = _get_output_ket(tape)
+
+    if len(tape.observables) == 1:
+        obs = qml.s_prod(cotangents[0], tape.observables[0])
+    else:
+        obs = qml.sum(*(qml.s_prod(cot, obs) for cot, obs in zip(cotangents, tape.observables)))
```

Comment:
perhaps we should change `qml.sum` to match `qml.prod` and accept only one operator, and when that happens, just return the operator itself? then we wouldn't have to do this if-else check

Reply:
Potentially could also use `qml.dot` here, `qml.dot(cotangents, tape.observables)`.